src/main/java/com/yelp/nrtsearch/server/luceneserver/search/collectors/MyTopSuggestDocsCollector.java

```diff
@@ -79,7 +93,119 @@ public TopSuggestDocs reduce(Collection<TopSuggestDocsCollector> collectors)
       for (TopSuggestDocsCollector collector : collectors) {
         topDocs[i++] = collector.get();
       }
-      return TopSuggestDocs.merge(numHitsToCollect, topDocs);
+
+      // create a map with docId for key and the best SuggestScoreDoc with the docId
+      // the "best" suggestScoreDoc is the one with the highest score
+      Map<Integer, TopSuggestDocs.SuggestScoreDoc> topSuggestScoreDocsUniqueByDoc = new HashMap<>();
+      for (TopSuggestDocs topSuggestDocs : topDocs) {
+        for (TopSuggestDocs.SuggestScoreDoc currentSearchSuggestDoc :
+            topSuggestDocs.scoreLookupDocs()) {
+          int docId = currentSearchSuggestDoc.doc;
+          if (topSuggestScoreDocsUniqueByDoc.containsKey(docId)) {
+            TopSuggestDocs.SuggestScoreDoc existingSearchSuggestDocForDocId =
+                topSuggestScoreDocsUniqueByDoc.get(docId);
+            if (existingSearchSuggestDocForDocId.score < currentSearchSuggestDoc.score) {
+              topSuggestScoreDocsUniqueByDoc.put(docId, currentSearchSuggestDoc);
+            }
+          } else {
+            topSuggestScoreDocsUniqueByDoc.put(docId, currentSearchSuggestDoc);
+          }
+        }
+      }
+
+      // retrieve the top this.numHitsToCollect suggestScoreDocs, sorting them by score
+      TopSuggestDocs.SuggestScoreDoc[] suggestScoreDocs =
+          topSuggestScoreDocsUniqueByDoc.values().stream()
+              .sorted((o1, o2) -> -1 * Float.compare(o1.score, o2.score)) // highest score first
+              .limit(this.numHitsToCollect)
+              .toArray(TopSuggestDocs.SuggestScoreDoc[]::new);
+
+      return new TopSuggestDocs(
+          new TotalHits(this.numHitsToCollect, TotalHits.Relation.EQUAL_TO), suggestScoreDocs);
+    }
+  }
+
+  public static class ExhaustiveTopSuggestDocsCollector extends TopSuggestDocsCollector {
+    private final SuggestScoreDocPriorityQueue priorityQueue;
+
+    public ExhaustiveTopSuggestDocsCollector(int num) {
+      super(num, false);
+      priorityQueue = new SuggestScoreDocPriorityQueue(num);
+    }
+
+    @Override
+    protected void doSetNextReader(LeafReaderContext context) {
+      docBase = context.docBase;
+      // clears priority queue
+      priorityQueue.getResults();
```

Comment:
I don't think you want to clear the queue here. This collector will process multiple leaves (segments) and you will want the queue to rank all the results together.

Reply:
resolved